cpp/src/gandiva/engine.cc

```diff
@@ -274,6 +277,27 @@ Status Engine::LoadPreCompiledIR() {
   return Status::OK();
 }
 
+Status Engine::LoadExternalPreCompiledIR() {
+  auto const& buffers = LLVMExternalIRStore::GetIRBuffers();
+  for (auto const& buffer : buffers) {
+    auto module_or_error = llvm::parseBitcodeFile(buffer->getMemBufferRef(), *context());
+    if (!module_or_error) {
+      std::string str;
+      llvm::raw_string_ostream stream(str);
+      stream << module_or_error.takeError();
+      return Status::CodeGenError("Failed to parse bitcode file, error: " + stream.str());
+    }
+    auto ir_module = std::move(module_or_error.get());
+
+    ARROW_RETURN_IF(llvm::verifyModule(*ir_module, &llvm::errs()),
+                    Status::CodeGenError("verify of IR Module failed"));
+    ARROW_RETURN_IF(llvm::Linker::linkModules(*module_, std::move(ir_module)),
+                    Status::CodeGenError("failed to link IR Modules"));
```

Comment:
Is there a way to get a richer error description from LLVM here? Saying that it failed without saying why is not useful to the caller...

Reply:
For `llvm::verifyModule`, the parameter `&llvm::errs()` will print the error into stdout. Now I change it to a `std::string` backed ` llvm::raw_string_ostream`, and return it as part of the `Status`.

For `llvm::Linker::linkModules`, I read the API doc (LLVM 14 in my env), and this API only returns a boolean value indicating if an error occurs, but I don't find any API to retrieve the detailed error message in this case, so I keep it as it is currently.